<?xml version="1.0"?>
<launch>
    <!-- 2011_09_26/2011_09_26_drive_0029_sync/testbag.bag -->
    <!-- [05] 2011_09_30/2011_09_30_drive_0018_sync/testbag.bag -->
    <node pkg="rosbag" type="play" name="rosbag_play" args="--clock -r 0.5 $(env HOME)/Downloads/kitti-raw/2011_09_30/2011_09_30_drive_0018_sync/testbag.bag"/>

    <!-- For Velodyne VLP-16 
    <param name="scan_line" value="16" />
    -->
    <!-- For Velodyne HDL-32 
    <param name="scan_line" value="32" />
    -->
    <!-- For Velodyne HDL-64 -->
    <param name="scan_line" value="64" />

    <!--- Sim Time -->
    <param name="/use_sim_time" value="true" />
    <param name="scan_period" value="0.1" /> 
    <param name="vertical_angle" type="double" value="2.0" />
    <param name="max_dis" type="double" value="90.0" />
    <param name="min_dis" type="double" value="3.0" />

    <!-- Height of Layser is 1.73m, so ground plane is on z=-1.73 -->
    <param name="threshold_height" value = "-1.3" />
    <!-- Width of traffic lane -->
    <param name="threshold_width" value = "15" />
    <param name="sdf_x_bound" value = "15" />
    <param name="sdf_y_bound" value = "30" />
    <param name="sdf_z_lower" value = "-2" />

    <param name="sdf_resolution" value = "1" />
    <param name="useFilteredGroundPoints" value="false" />
    <!-- -->
    <node pkg="floam" type="floam_laser_processing_node" name="floam_laser_processing_node" output="screen"/>
    <node pkg="floam" type="floam_odom_estimation_node" name="floam_odom_estimation_node" output="screen"/>
    
    <!-- 以10Hz发布从坐标系 /world 到坐标系 /map的静态坐标变换 000000意味着这两个坐标系是重叠的-->
    <node pkg="tf" type="static_transform_publisher" name="world2map_tf"  args="0 0 0 0 0 0 /world /map 10" />
    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find floam)/rviz/floam.rviz" />
    </group>

    <!-- 用于记录轨迹。默认话题类型为'nav_msgs/Path'。默认发布话题为'/trajectory'
        /ns：命名空间
        /target_frame_name： tf中的目标坐标系
        /source_frame_name： tf中的源坐标系
        /trajectory_update_rate：被保存的轨迹trajectory的更新频率
        /trajectory_publish_rate：轨迹的发布频率，默认发布到trajectory话题
    -->
    <!-- 真值轨迹发布到topic: /gt/trajectory 
        velodyne是rosbag中的topic: velodyne_points的frame_id
    -->
  	<node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="gt" >
        <param name="/target_frame_name" value="world" />
        <param name="/source_frame_name" value="velodyne" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>
    <!-- 算法轨迹发布到topic: /base_link/trajectory 
        base_link是机器人的frame_id
    -->
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="base_link" >
        <param name="/target_frame_name" value="world" />
        <param name="/source_frame_name" value="base_link" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>

    <arg name="ifsave" default="false"/>
    <group if="$(arg ifsave)">
        <param name="saved_path" value="/home/yuxuanzhao/ros_workspace/floam_ws/src/floam/result/evo_result/kitti_05_sdf.txt" />
        <node pkg="floam" type="save_pose_node" name="save_pose_node" output="screen"/>
    </group>
</launch>
